### ISSUE TITLE
Bug in inflight message restore from datastore

### ISSUE THREAD

**Issue by s9-hfe** [mochi-mqtt/server#406]
First, let me thank you for your effort in creating this excellent piece of software.

I noticed an issue in the following snippet where you are restoring messages from the datastore. You are assigning the messages to the Origin instead of the Client who should actually receive the message.

Another point: if the client can't be found, shouldn't the message be dropped in the datastore? The same applies to expired messages.

```
// loadInflight restores inflight messages from the datastore.
func (s *Server) loadInflight(v []storage.Message) {
	for _, msg := range v {
		if client, ok := s.Clients.Get(msg.Origin); ok {  // <--- Why Origin?
			client.State.Inflight.Set(msg.ToPacket())
		}
	}
}
```

BR
HFE

**Comment by werbenhu** [comment]
> Another point: if the client can't be found, shouldn't the message be dropped in the datastore? The same applies to expired messages.

You can take a look at loadClients(). If a client has expired and needs to be deleted, the function will clean up the inflight messages.

**Comment by werbenhu** [comment]
@s9-hfe In MQTT protocol with QoS 2, the sending and receiving processes are indeed separate but follow the same four-step handshake. 
QoS 2  complete flow:

**Sending side (Client A to Server)**
1. Client A -> Server: PUBLISH
2. Server -> Client A: PUBREC
3. Client A -> Server: PUBREL
4. Server -> Client A: PUBCOMP

**Receiving side (Server to Client B)**

5. Server -> Client B: PUBLISH
6. Client B -> Server: PUBREC
7. Server -> Client B: PUBREL
8. Client B -> Server: PUBCOMP

The sending and receiving processes are two independent flows, each with its own four-step handshake mechanism.Each flow ensures exactly-once message delivery.

So, there's no issue here.

**Comment by s9-hfe** [comment]
Try the following:

- Run MQTT with persistence: go run examples/persistence/badger/main.go (but remove line "defer os.RemoveAll(badgerPath)" before)
- Client A subscribes topic "test" with QOS1 and clean-session disabled
- Client B publishes to topic "test"
- Client A receives the message (ok so far)
- Stop Client A
- Client B publishes again to topic "test". Client A is not running anymore, so the message gets persisted (but with the Origin-ID which is Client B and not Client A).
**- IMPORTANT: restart the MQTT server now**
- Start Client A again
- Client A does not get the messages redelivered (because they were not persisted with the ID of Client A)

See https://github.com/mochi-mqtt/server/pull/427

**Comment by thedevop** [comment]
You're correct @s9-hfe! Although the messages are persisted with the ID that contains the clientID, however, it is not used when restoring the message. The msg.Original is used instead, which will always be the publisher client ID and not the subscriber client ID. The PR you provided looks reasonable.

**Comment by werbenhu** [comment]
@s9-hfe I reviewed your PR, and it seems you are right. The inflight should be stored in the receiver's state.

**Comment by mochi-co** [comment]
@werbenhu @thedevop @s9-hfe I will try to go through all the open PRs and merge where necessary in the next day or so :)

**Comment by thedevop** [comment]
Fixed with PR #427 in release v2.6.6.
